control-plane/agents/src/bin/core/node/specs.rs

```diff
@@ -144,135 +114,90 @@ impl ResourceSpecsLocked {
             })
             .collect()
     }
+}
 
-    /// Apply the drain label to trigger draining.
-    /// Return the NodeSpec.
-    pub(crate) async fn drain_node(
-        &self,
-        registry: &Registry,
-        node_id: &NodeId,
-        label: String,
-    ) -> Result<NodeSpec, SvcError> {
-        let node = self.node_rsc(node_id)?;
-        let drained_node_spec = {
-            let mut locked_node = node.lock();
-            // Do not allow the same label to be applied more than once.
-            if locked_node.has_cordon_label(&label) {
-                return Err(SvcError::CordonLabel {
-                    node_id: node_id.to_string(),
-                    label,
-                });
-            }
-            locked_node.set_drain(label);
-            locked_node.clone()
-        };
-        registry.store_obj(&drained_node_spec).await?;
-        Ok(drained_node_spec)
-    }
+#[async_trait::async_trait]
+impl GuardedOperationsHelper for OperationGuardArc<NodeSpec> {
+    type Create = NodeSpec;
+    type Owners = ();
+    type Status = ();
+    type State = NodeSpec;
+    type UpdateOp = NodeOperation;
+    type Inner = NodeSpec;
 
-    /// Set the NodeSpec to the drained state.
-    pub(crate) async fn set_node_drained(
-        &self,
-        registry: &Registry,
-        node_id: &NodeId,
-    ) -> Result<NodeSpec, SvcError> {
-        let node = self.node_rsc(node_id)?;
-        let drained_node_spec = {
-            let mut locked_node = node.lock();
-            locked_node.set_drained();
-            locked_node.clone()
-        };
-        registry.store_obj(&drained_node_spec).await?;
-        Ok(drained_node_spec)
+    fn remove_spec(&self, _registry: &Registry) {
+        unimplemented!();
     }
+}
```

Comment:
Can we not skip this impl?

Reply:
You mean take out this trait implementation?
AFAIUI OperationGuardArc<NodeSpec> needs this trait for the default implementations of start_update() and complete_update().
There is no default implementation of remove_spec() so the code won't compile without an implementation for OperationGuardArc<NodeSpec>, but no code calls this function.

Reply:
Ahh okay.